### ISSUE TITLE
Validate ndarray dtype in C++ wrapper.

### ISSUE THREAD

**Issue by ailzhang** [taichi-dev/taichi#7276]
In deployment scenarios, we may accidentally pass in an ndarray with wrong dtype compared to the one required by the compiled kernel, we should throw proper error to users when it happens.

**Comment by ailzhang** [comment]
Also we don't check number of args matches compiled kernel so we get segfaults like https://github.com/taichi-dev/taichi-nerf/issues/18, maybe also good to validate that and throw a proper error :D

**Comment by lgyStoic** [comment]
right! I encounter this issue when I using u8 ndarray as passed value in JIT whih parameter is u32 ndarray, taichi JIT convert the type automatic. But in AOT mode, I pass u8 ndarray to module, the whole algorithm is not right ! I think it's neccessory  adding the type checker for taichi AOT
